pkg/endpoint/endpoint.go

```diff
@@ -1506,19 +1520,44 @@ func (e *Endpoint) Update(owner Owner, opts models.ConfigurationMap) error {
 	}
 
 	if needToRegenerate {
-		stateTransitionSucceeded := e.SetStateLocked(StateWaitingToRegenerate, reason)
-
-		e.Mutex.Unlock()
-		ctCleaned.Wait()
-
-		if stateTransitionSucceeded {
-			e.Regenerate(owner, reason)
-			return nil
+		scopedLog.Debug("need to regenerate endpoint; checking state before" +
+			" attempting to regenerate")
+
+		// Default timeout for PATCH /endpoint/{id}/config is 30 seconds, so put
+		// timeout in this function a bit below that timeout. If the timeout
+		// for clients in API is below this value, they will get a message containing
+		// "context deadline exceeded".
+		stateChangeTimeout := time.Duration(25 * time.Second)
+
+		// Check up until stateChangeTimeout seconds for endpoint state before
+		// trying to update configuration.
+		timeout := time.After(stateChangeTimeout)
+
+		// Check for endpoint state every second.
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				// Check endpoint state before attempting configuration update because
+				// configuration updates can only be applied when the endpoint is in
+				// specific states. See GH-3058.
+				stateTransitionSucceeded := e.SetStateLocked(StateWaitingToRegenerate, reason)
+				if stateTransitionSucceeded {
+					e.Mutex.Unlock()
+					ctCleaned.Wait()
+					e.Regenerate(owner, reason)
+					return nil
+				}
+			case <-timeout:
+				scopedLog.Warningf("timed out waiting for endpoint state to change")
+				e.Mutex.Unlock()
+				ctCleaned.Wait()
+				return UpdateStateChangeError{fmt.Sprintf("unable to regenerate endpoint program because state transition to %s was unsuccessful; check `cilium endpoint log %d` for more information", StateWaitingToRegenerate, e.ID)}
+			}
 		}
 
-		// FIXME: GH-3058: We need to queue up a regeneration
```

Comment:
Why not use `pkg/serializer`?

Reply:
This is a short-term fix. See the discussion: https://github.com/cilium/cilium/issues/3058 . I agree that something like `pkg/serializer` is the way to go, but adding something like it right now seems risky compared to this change.

Reply:
If it's a short-term fix then leave the `FIXME` please

Reply:
Good point, will do.

Reply:
Updated to point to issue which is for discussing how to queue up builds per endpoint.